Add methods to draw ellipses

Cykyrios opened the pull request:
Closes godotengine/godot-proposals#8461.

This simply modifies the formulas for `draw_circle` and `draw_arc` to generalize to ellipses and elliptical arcs.

Calinou reviewed (commented):
I suggest modifying `draw_circle()`'s description to point to `draw_ellipse()` and vice versa. The same goes for `draw_arc()` which should refer to `draw_ellipse_arc()` too.

Cykyrios commented:
Rebased, there were some changes in `draw_circle` that I replicated for ellipses.
Side note: while updating the class documentation, I noticed a typo in several `draw_*` methods, should I include this here as well? (only `draw_circle` is relevant to this PR)

Calinou commented:
> Side note: while updating the class documentation, I noticed a typo in several draw_* methods, should I include this here as well? (only draw_circle is relevant to this PR)

I'd open a separate PR as it's an independent change.

TheEskir commented:
Apologies in advance if this is frowned upon here but - bump? Can someone who qualifies review+merge this _please?_

This is the way draw_circle should have been implemented from the start.
Not having anything to draw_ellipse is a pretty big oversight.

Using GDScript and a loop to get points for a Vector2 Array that is then drawn with, for example, draw_multiline kills performance like hell. Not ideal.

Using draw_set_transform on a circle to scale height only works for filled-in circles, for line-only circles without infill, it distorts the line thickness due to the scaling. Also not ideal.

Using an ellipse sprite seems like a dumb workaround for what I want to do (and also suffers from the squishing/stretching on set_transform scaling)

Calinou reviewed (approved):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/f0f5319b0b5b56db9f8023ac4132088df9e86b9e), it works as expected.

**Testing project:** [test_draw_ellipse.zip](https://github.com/user-attachments/files/18693048/test_draw_ellipse.zip)

From left to right and top to bottom - Circle, ellipse, arc, ellipse arc (arcs with 6 points only):

<img width="69" alt="Screenshot 2025-02-06 at 17 16 39" src="https://github.com/user-attachments/assets/db794c6e-686a-467a-8645-decbe0a81c04" />

Something strange is that circle always draws as filled, while ellipse works as intended. This occurs both in Forward+ and Compatibility.

Cykyrios commented:
Sorry about that, you're right that I broke circle fill... I was making a simple call to
```c++
RenderingServer::get_singleton()->canvas_item_add_ellipse()
```
instead of simply calling `draw_ellipse()`. Fixed and rebased.
![circle_ellipse](https://github.com/user-attachments/assets/b2896a15-12b7-4107-a562-c3f3e09ff507)

Calinou commented:
Works great now:

![image](https://github.com/user-attachments/assets/889a2575-9efd-4447-b95a-c05ad8692600)

Cykyrios commented:
Rebased.

Repiteo commented:
Needs another rebase

Cykyrios committed:
Add methods to draw ellipses

Cykyrios commented:
Rebased and added `ERR_DRAW_GUARD` (this macro and `ERR_THREAD_GUARD` are now called twice when drawing circles and circular arcs).

AThousandShips commented on scene/main/canvas_item.cpp:
```diff
@@ -845,9 +854,9 @@ void CanvasItem::draw_circle(const Point2 &p_pos, real_t p_radius, const Color &
-	} else if (p_width >= 2.0 * p_radius) {
-		RenderingServer::get_singleton()->canvas_item_add_circle(canvas_item, p_pos, p_radius + 0.5 * p_width, p_color, p_antialiased);
+		RenderingServer::get_singleton()->canvas_item_add_ellipse(canvas_item, p_pos, p_major, p_minor, p_color, p_antialiased);
+	} else if (p_width >= 2.0 * fmax(p_major, p_minor)) {
```
```suggestion
	} else if (p_width >= 2.0 * MAX(p_major, p_minor)) {
```

Cykyrios commented:
Applied suggested change, and also removed `"draw_circle()"` from the width warning, as it also applies to `draw_ellipse()`, and the method appears in the stack trace.
